src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1249,32 +1257,42 @@ private static MethodHandle generate(Lookup lookup, MethodType args, String[] co
                 return handle.bindTo(concat1);
             }
 
-            var weakConstructorHandle = CACHE.get(concatArgs);
-            if (weakConstructorHandle != null) {
-                MethodHandlePair handlePair = weakConstructorHandle.get();
-                if (handlePair != null) {
-                    try {
-                        var instance = handlePair.constructor.invoke(constants);
-                        return handlePair.concatenator.bindTo(instance);
-                    } catch (Throwable e) {
-                        throw new StringConcatException("Exception while utilizing the hidden class", e);
+            boolean forceInline  = concatArgs.parameterCount() <  FORCE_INLINE_THRESHOLD;
+            boolean staticConcat = concatArgs.parameterCount() >= CACHE_THRESHOLD;
+
+            if (!staticConcat) {
+                var weakConstructorHandle = CACHE.get(concatArgs);
+                if (weakConstructorHandle != null) {
+                    MethodHandlePair handlePair = weakConstructorHandle.get();
+                    if (handlePair != null) {
+                        try {
+                            var instance = handlePair.constructor.invoke(constants);
+                            return handlePair.concatenator.bindTo(instance);
+                        } catch (Throwable e) {
+                            throw new StringConcatException("Exception while utilizing the hidden class", e);
+                        }
                     }
                 }
             }
+
             MethodTypeDesc lengthArgs  = lengthArgs(concatArgs),
                            coderArgs   = coderArgsIfMaybeUTF16(concatArgs),
-                           prependArgs = prependArgs(concatArgs);
+                           prependArgs = prependArgs(concatArgs, staticConcat);
 
             byte[] classBytes = ClassFile.of().build(CD_CONCAT,
                     new Consumer<ClassBuilder>() {
-                        final boolean forceInline = concatArgs.parameterCount() < FORCE_INLINE_THRESHOLD;
-
                         @Override
                         public void accept(ClassBuilder clb) {
-                            clb.withSuperclass(CD_StringConcatBase)
-                                .withFlags(ACC_FINAL | ACC_SUPER | ACC_SYNTHETIC)
-                                .withMethodBody(INIT_NAME, MTD_INIT, 0, CONSTRUCTOR_BUILDER)
-                                .withMethod("length",
+                            if (staticConcat) {
+                                clb.withSuperclass(CD_Object)
+                                   .withFlags(ACC_FINAL | ACC_SUPER | ACC_SYNTHETIC);
```

Comment:
According to #19517, project lilliput wants utility classes to be declared `abstract` instead of `final` so their pointers won't be encodable in object headers

Reply:
Do I need to declare it as ACC_INTERFACE? Many utility classes do this.

Reply:
interface has extra restrictions that can fail class validation, such as fields must be public static final. So I recommend just using abstract class.

Reply:
Thanks, I learned something new!